django/views/decorators/http.py

```diff
@@ -40,9 +40,26 @@ def inner(request, *args, **kwargs):
                     request=request,
                 )
                 return response
-            return func(request, *args, **kwargs)
 
-        return inner
+        async def _view_wrapper_async(request, *args, **kwargs):
+            invalid_response = _is_method_not_valid(request)
+            return (
+                await func(request, *args, **kwargs)
+                if not invalid_response
+                else invalid_response
+            )
+
+        def _view_wrapper(request, *args, **kwargs):
+            invalid_response = _is_method_not_valid(request)
+            return (
+                func(request, *args, **kwargs)
+                if not invalid_response
+                else invalid_response
+            )
+
+        return wraps(func)(
+            _view_wrapper_async if iscoroutinefunction(func) else _view_wrapper
+        )
```

Comment:
This approach is "DRY" but this "cleanliness" comes with relatively large runtime penalties: three functions are generated for every call to the decorator instead of one, and on every request there is an extra function call to `_is_method_not_valid`.

I'd prefer if we did the dumb duplicative thing and made the appropriate decorator function at apply time:

```python
def require_http_methods(request_method_list):
    """
    Decorator to make a view only accept particular request methods.  Usage::

        @require_http_methods(["GET", "POST"])
        def my_view(request):
            # I can assume now that only GET or POST requests make it this far
            # ...

    Note that request methods should be in uppercase.
    """

    def decorator(func):
        if iscoroutinefunction(func):
            
            @wraps(func)
            async def inner(request, *args, **kwargs):
                if request.method not in request_method_list:
                    response = HttpResponseNotAllowed(request_method_list)
                    log_response(
                        "Method Not Allowed (%s): %s",
                        request.method,
                        request.path,
                        response=response,
                        request=request,
                    )
                    return response
                return await func(request, *args, **kwargs)

        else:
          
            @wraps(func)
            def inner(request, *args, **kwargs):
                if request.method not in request_method_list:
                    response = HttpResponseNotAllowed(request_method_list)
                    log_response(
                        "Method Not Allowed (%s): %s",
                        request.method,
                        request.path,
                        response=response,
                        request=request,
                    )
                    return response
                return func(request, *args, **kwargs)
          
          return inner

    return decorator
```

ditto below